Refactored fit() interface

karl-richter opened the pull request:
## :microscope: Background

- We want to provide a user with the ability to activate / deactivate some of the non-required features during training. This includes checkpointing, logging of progress via the progress bar as well as the calculation of metrics during training.
- In the initial version of NeuralProphet, the `minimal` used a separate function to `_train()` called `_train_minimal()` that did not contain all of the above mentioned features and was "minimal".
- With the transition to Lightning, the training procedure is abstracted into the model and a lot simpler. Thus, we removed the `_train()` function and added `minimal` as a parameter to it.
- With the performance optimization of Lightning, the demand raised to control the training features individually. Thus, this PR supports controlling this behavior more granularly by using the `progress`, `metrics` and `checkpointing` parameter. The `minimal` mode further persists, but only remains a shortcut to set the other parameters to false.

## :crystal_ball: Key changes

- Added a deprecation warning for the `collect_metrics` parameter in `__init__` since this information can only be provided in the fit method in the future
- Added the `metrics` and `checkpointing` parameter to the `fit()` and subsequent functions
- Small refactorings and clean ups

## :clipboard: Review Checklist
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, added docstrings and data types to function definitions.
- [x] I have added pytests to check whether my feature / fix works.

Please make sure to follow our best practices in the [Contributing guidelines](https://github.com/ourownstory/neural_prophet/blob/main/CONTRIBUTING.md).

karl-richter commented:
Hey @ourownstory @noxan, this PR would bring some interface changes (remove minimal, add metrics and checkpointing). Should we provide the metrics and checkpointing arg in NeuralProphet or in the fit() function?

noxan commented:
@karl-richter I like your approach, also agree with moving the metrics and checkpoints to where they actually matter (in the fit function) - would also make it very clear if people disabled metrics that they cannot expect that as return value.

karl-richter commented:
**Interface Change**
```python
# 0.4.2
def fit(
        self,
        df,
        freq="auto",
        validation_df=None,
        progress="bar",
        minimal=False
)

# 0.5.0
def fit(
        self,
        df: pd.DataFrame,
        freq: str = "auto",
        validation_df: pd.DataFrame = None,
        epochs: int = None,
        batch_size: int = None,
        learning_rate: float = None,
        early_stopping: bool = False,
        minimal: bool = False,
        metrics: bool = None,
        progress="bar",
        checkpointing: bool = False,
        continue_training: bool = False,
    )
```

karl-richter commented:
Added preparation for distributed training on request by @judussoari. This includes the `num_workers` param in `fit()`

noxan commented:
@karl-richter Many thanks for the great work. I'm still working on the review. For the next pull requests it would be great to split this into several smaller ones, as it is hard to review everything at once and especially to keep track which changes relate to what...

1. Replace minimal mode by checkpoints and metrics
2. Refactor fit method to allow continue training with other parameters
3. Prepare support for `num_workers`
4. A changed checkpoints implementation
5. Moving the progress bar implementation
6. ...

Happy to have a chat with you to explain this in detail, I'll try to conclude the review soon :)

noxan commented on neuralprophet/forecaster.py:
```diff
@@ -654,7 +656,22 @@ def add_seasonality(self, name, period, fourier_order):
+        learning_rate: float = None,
+        early_stopping: bool = False,
+        minimal: bool = False,
+        metrics: bool = None,
```
Or is it a special way of using the previously defined metrics? I mean that is a pretty confusing way if people want to explicitly define this 🤔

noxan commented on neuralprophet/forecaster.py:
```diff
@@ -726,18 +792,30 @@ def fit(self, df, freq="auto", validation_df=None, progress="bar", minimal=False
+            metrics_df = self._train(
+                df,
+                df_val=df_val,
+                progress=bool(progress),
```
Same as with metrics comment above, we should have clear variable naming if they contain different values in different places

noxan commented on neuralprophet/utils.py:
```diff
@@ -833,6 +837,6 @@ def _update_n(self, bar, value: int) -> None:
     # TODO: Disabling sampler_ddp brings a good speedup in performance, however, check whether this is a good idea
     # https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#replace-sampler-ddp
-    config["replace_sampler_ddp"] = False
+    # config["replace_sampler_ddp"] = False
```
Is this an intentional change?

noxan commented:
@karl-richter Overall a great set of changes. Might be best to have a short call on this. We should consider extracting changes into separate PRs and then review and merge them individually.

karl-richter commented on neuralprophet/forecaster.py:
```diff
@@ -319,7 +317,6 @@ def __init__(
         ar_reg: Optional[float] = None,
         learning_rate: Optional[float] = None,
         epochs: Optional[int] = None,
-        early_stopping: bool = False,
```
Was never released in a public version, thus this should be fine

karl-richter commented on neuralprophet/forecaster.py:
```diff
@@ -654,7 +656,22 @@ def add_seasonality(self, name, period, fourier_order):
+        learning_rate: float = None,
+        early_stopping: bool = False,
+        minimal: bool = False,
+        metrics: bool = None,
```
```suggestion
        metrics: np_types.CollectMetricsMode = None,
```

karl-richter commented on neuralprophet/forecaster.py:
```diff
@@ -701,6 +758,15 @@ def fit(self, df, freq="auto", validation_df=None, progress="bar", minimal=False
+            log.warning("Progress plot requires metrics to be enabled. Enabling the default metrics.")
+            metrics = metrics.get_metrics(True)
+
+        if minimal:
+            checkpointing = False
+            self.metrics = False
+            progress = None
```
Consider removing the minimal parameter or add a warning in case any other configuration (eg. checkpointing etc) is provided together with minimal.

karl-richter commented on neuralprophet/time_net.py:
```diff
@@ -203,7 +199,8 @@ def __init__(
 
         # Metrics Config
-        if not minimal:
+        self.collect_metrics = bool(metrics)  # yields True if metrics is not an empty dictionary
+        if self.collect_metrics:
```
#1052

karl-richter commented:
@noxan I adressed most of our comments:
- renamed `metrics.py` to `utils_metrics.py`
- renamed `collect_metrics` to `metrics_enabled` in `time_net.py`
- renamed `checkpointing` to `checkpointing_enabled` (and similar for metrics, progress_bar etc.)
- fixed non-pythonic return in `_train()` (no more if else for return)

noxan reviewed (approved):
LGTM
